### PR TITLE
Fix triangle splitting crash where FP error causes binning differences

### PR DESCRIPTION
**Pull request by jmsaunde**
Ultimately the root problem here is that when we do the binning stage we use a
different method to test the first stage and the second. In the first stage we
use bin() to get an integer bin associated with a point in space. In the second
we get the floating point division point on the axis we have chosen to bin on
and do FP comparisons instead. Under most circumstances this is fine.

Occasionally these binning differences will result in unprofitable splits
happening but still building a usable tree.

In the test case I've given, the unprofitable splits we erroneously choose
fill up the split limit and we don't end up splitting the triangles the
previous stage had intended to choose. As a result, we can't split those
triangles. Then when we do the real bifurcation, everything is on the same
side. This resulted in an assertion failure in production (1).

(1) We never noticed that the RelWithDebInfo build in MSVC still has assertions
enabled. I can provide another change disabling these since I think it's
unintentional.

### CODE REVIEW

File: kernels/builders/heuristic_spatial_array.h [hunk 1]
```diff
@@ -283,11 +283,9 @@ namespace embree
 
                 if (likely(splits <= 1)) continue; /* todo: does this ever happen ? */
 
-                //int bin0 = split.mapping.bin(prims0[i].lower)[split.dim];
```
Comment (jmsaunde) [inline kernels/builders/heuristic_spatial_array.h:286]: I'd like to understand why this change was made a couple of years ago. Ultimately it was the source of our issue here.

File: kernels/common/state.cpp [hunk 2]
```diff
@@ -428,6 +430,9 @@ namespace embree
       else if (tok == Token::Id("max_spatial_split_replications") && cin->trySymbol("="))
         max_spatial_split_replications = cin->get().Float();
 
+      else if (tok == Token::Id("max_triagles_per_leaf") && cin->trySymbol("="))
```
Comment (stanard) [inline kernels/common/state.cpp:433]: Misspelled

File: kernels/builders/heuristic_spatial_array.h (hunk 1 above)
Comment (jmsaunde) [reply kernels/builders/heuristic_spatial_array.h:286]: @cbenthin @freibold re https://github.com/RenderKit/embree/commit/190a0dc5f757f5e93a94851938679e089bc449f4

**Comment by freibold** [comment]
Hi! Thanks for this PR and especially for the small test! The changes look good. However, I've tried to verify that the test fails in the current Embree devel branch but it's always passing. It also passes in our internal CI running on various machines with many different configs. Can you double check that the test is really a reproducer that triggers asserts during the build (i.e, without the other changes you made). What config (VS version, OS version, CMake options, build type, ....) are you using?

**Comment by jmsaunde** [comment]
> Hi! Thanks for this PR and especially for the small test! The changes look good. However, I've tried to verify that the test fails in the current Embree devel branch but it's always passing. It also passes in our internal CI running on various machines with many different configs. Can you double check that the test is really a reproducer that triggers asserts during the build (i.e, without the other changes you made). What config (VS version, OS version, CMake options, build type, ....) are you using?

Thanks for reviewing!

It was much easier to reproduce the issue by adding the feature in the first commit: supporting setting the max leaf size. The test needs and uses this feature to reproduce the issue. I believe an analogous test should exist without adding this but finding it would be difficult.

I'm running on MSVC2022/Win11. I'm off work today, but can follow up tomorrow with the other build options that I used. There were actually several CMake configuration settings we use internally so it's a good point.

**Comment by freibold** [comment]
Thanks, I was missing that the test uses the max_triangles_per_leaf setting and could now test the PR properly. I've merged this into our internal devel branch and it will be available publicly very soon.
